src/update-helper/update-helper.c

```
@@ -0,0 +1,946 @@
+                        _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+                        r = bus_call_method(
+                                        bus,
+                                        bus_systemd_mgr,
+                                        "StopUnit",
+                                        &error,
+                                        &reply,
+                                        "ss", *unit, "replace");
+                        if (r < 0) {
+                                log_full_errno(arg_quiet ? LOG_DEBUG : LOG_WARNING, r,
+                                               "Failed to stop unit %s, ignoring: %s",
+                                               *unit, bus_error_message(&error, r));
+                                continue;
+                        }
+
+                        const char *path;
+                        r = sd_bus_message_read(reply, "o", &path);
+                        if (r < 0)
+                                return bus_log_parse_error(r);
+
+                        r = bus_wait_for_jobs_add(w, path);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to watch job for %s: %m", *unit);
+                }
+
+                r = bus_wait_for_jobs(w, BUS_WAIT_JOBS_LOG_ERROR, /* extra_args= */ NULL);
+                if (r < 0)
+                        return r;
+        } else {
+                _cleanup_strv_free_ char **users = NULL;
+
+                r = list_units(bus, STRV_MAKE("user@*.service"), &users);
+                if (r < 0)
+                        return r;
+
+                STRV_FOREACH(user, users) {
+                        _cleanup_(sd_bus_flush_close_unrefp) sd_bus *user_bus = NULL;
+
+                        r = bus_connect_user_unit(*user, &user_bus);
+                        if (r < 0)
+                                continue;
```

Maybe, better to not ignore all errors in bus_connect_user_unit(), but make some errors critical, and others non-critical. Then, return 1 when connected, and 0 on non-critical errors.